### PR TITLE
SRVKP-6439 : Implement s3 provide for cache step action

### PR DESCRIPTION
**Pull request by pramodbindal**

### CODE REVIEW

File: internal/fetch/fetch.go [hunk 1]
```diff
@@ -27,7 +28,7 @@ func Fetch(ctx context.Context, hash, target, folder string, insecure bool) erro
 	case "oci":
 		return oci.Fetch(ctx, hash, newTarget, folder, insecure)
 	case "s3":
-		return fmt.Errorf("s3 schema not (yet) supported: %s", target)
+		return s3.DownloadFromS3(ctx, hash, newTarget, folder)
```
Comment (vdemeester) [inline internal/fetch/fetch.go:31]: For the other implementation I used `Fetch` and `Upload` for the function name (to be consistent). Could we do the same here ?

File: internal/provider/common_utils.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package provider
```
Comment (vdemeester) [inline internal/provider/common_utils.go:1]: We could use a `tar` package of something ? (`tar.File`  ?). `common_utils.go` doesn't really say anything.

File: internal/provider/s3/s3.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
+package s3
+
+import (
```
Comment (vdemeester) [inline internal/provider/s3/s3.go:3]: nit: the imports needs to be organized (as they are in other files).

File: internal/upload/upload.go [hunk 4]
```diff
@@ -20,7 +21,7 @@ func Upload(ctx context.Context, hash, target, folder string, insecure bool) err
 	case "oci":
 		return oci.Upload(ctx, hash, newTarget, folder, insecure)
 	case "s3":
-		return fmt.Errorf("s3 schema not (yet) supported: %s", target)
+		return s3.UploadToS3(ctx, hash, newTarget, folder)
```
Comment (vdemeester) [inline internal/upload/upload.go:24]: Same as above, can we use `Upload` as the method name ?

File: internal/provider/common_utils.go (hunk 2 above)
Comment (pramodbindal) [reply internal/provider/common_utils.go:1]: Refactored this

File: internal/fetch/fetch.go [hunk 5]
```diff
@@ -22,14 +25,25 @@ func Fetch(ctx context.Context, hash, target, folder string, insecure bool) erro
...
+		file, _ := os.CreateTemp("", "cache.tar")
+		if err := provider.Providers[u.Scheme].Fetch(ctx, source, file.Name()); err != nil {
+			return err
+		}
+		f, err := os.Open(file.Name())
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		return extract.Archive(ctx, f, folder, nil)
```
Comment (vdemeester) [inline internal/fetch/fetch.go:43]: I think this one should look as simple as the other one, like `s3.Fetch` (and "hide" implementation detail of the tarball, …)

File: internal/utils/file-utils.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package utils
```
Comment (vdemeester) [inline internal/utils/file-utils.go:1]: This should have another name. `utils` package are very badly name (utils doesn't mean anything — see https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common).

File: internal/upload/upload.go [hunk 7]
```diff
@@ -20,7 +24,17 @@ func Upload(ctx context.Context, hash, target, folder string, insecure bool) err
...
+
+		tarFile, err := os.CreateTemp("", "cache.tar")
+		if err != nil {
+			return err
+		}
+		if err := utils.Tarit(folder, tarFile.Name()); err != nil {
+			return err
+		}
+		defer os.Remove(tarFile.Name())
+		return provider.Providers[u.Scheme].Upload(ctx, newTarget, tarFile.Name())
```
Comment (vdemeester) [inline internal/upload/upload.go:37]: Same as above.

**Review by vdemeester — COMMENTED** [review]
- On `dev/step-action/*`, I think we should do something different than we do today. As it is in this PR, we have duplicated version of the stepaction, making it very error prone on updates (we update `dev/step-action/…` but not the other, …). We should have only one "real" copy (with `ko://` most likely) and one thing that is generated from the other (either copied, or released, or …)
- I would rather not have the `Provide` interface, and have a `s3` package that holds the code of the s3 implementation. As is, it doesn't provide much, it also adds all the `aws-sdk-go` dependency to the `provider` package (whigh might be an issue).

File: internal/fetch/fetch.go (hunk 5 above)
Comment (pramodbindal) [reply internal/fetch/fetch.go:43]: I am trying to keep a common implementation for all providers.
Once we do it for S3. we can remove all other conditions and  provider.Providers[u.Scheme] will get the correct provider where we can call update.

and creating/extracting tarball is common task for all providers so I wanted to have that here only

Comment (pramodbindal) [reply internal/fetch/fetch.go:43]: final implementation should look like  this

`func Upload(ctx context.Context, hash, target, folder string, insecure bool) error {
	u, err := url.Parse(target)
	if err != nil {
		return err
	}
	newTarget := strings.TrimPrefix(target, u.Scheme+"://")
	newTarget = strings.ReplaceAll(newTarget, "{{hash}}", hash)
	tarFile, err := os.CreateTemp("", "cache.tar")
	if err != nil {
		return err
	}
	if err := fileutils.Tarit(folder, tarFile.Name()); err != nil {
		return err
	}
	defer os.Remove(tarFile.Name())
	providerImpl := provider.Providers[u.Scheme]
	if providerImpl != nil {
		return fmt.Errorf("unknown schema: %s", target)
	}
	return providerImpl.Upload(ctx, newTarget, tarFile.Name())

}`

Comment (vdemeester) [reply internal/fetch/fetch.go:43]: > and creating/extracting tarball is common task for all providers so I wanted to have that here only

Well, it's not for the `oci` provider (or at least not in the same way).

Comment (pramodbindal) [reply internal/fetch/fetch.go:43]: > > and creating/extracting tarball is common task for all providers so I wanted to have that here only
> 
> Well, it's not for the `oci` provider (or at least not in the same way).

I have copied the tarit and extract code from oci only . so ideally this should be same.

**Comment by pramodbindal** [comment]
> * On `dev/step-action/*`, I think we should do something different than we do today. As it is in this PR, we have duplicated version of the stepaction, making it very error prone on updates (we update `dev/step-action/…` but not the other, …). We should have only one "real" copy (with `ko://` most likely) and one thing that is generated from the other (either copied, or released, or …)

**AGREE**

**Review by vdemeester — APPROVED** [review]

**Pull request merged**